src/db/lmdb_data_source_test.cc

```diff
@@ -0,0 +1,96 @@
+#include "src/db/lmdb_data_source.h"
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+namespace db {
+
+namespace {
+
+TEST(LMDBDataSourceTest, ConnectSuccessValidPath) {
+  EXPECT_NO_THROW(LMDBDataSource db_source("/tmp"));
+}
+
+TEST(LMDBDataSourceTest, ConnectFailWithInvalidPath) {
+  EXPECT_THROW(LMDBDataSource db_source("invalid db path"),
+               lmdb::runtime_error);
+}
+
+TEST(LMDBDataSourceTest, TransactionOps) {
+  LMDBDataSource db_source("/tmp");
+
+  // Test: Begin then Commit both succeed.
+  absl::Status status = db_source.Begin();
+  EXPECT_EQ(status.code(), absl::StatusCode::kOk) << status.ToString();
+
+  status = db_source.Commit();
+  EXPECT_EQ(status.code(), absl::StatusCode::kOk) << status.ToString();
+
+  // Test: Commit or Abort after a Commit returns FailedPrecondition error.
+  status = db_source.Commit();
+  EXPECT_EQ(status.code(), absl::StatusCode::kFailedPrecondition)
+      << status.ToString();
+
+  status = db_source.Abort();
+  EXPECT_EQ(status.code(), absl::StatusCode::kFailedPrecondition)
+      << status.ToString();
+}
+
+TEST(LMDBDataSourceTest, SingleTxnPutThenGet) {
+  LMDBDataSource db_source("/tmp");
+
+  const std::string key = "key";
+  const int64_t put_value = 123;
+
+  // Test: Put value succeed.
+  absl::Status status = db_source.Begin();
+  ASSERT_EQ(status.code(), absl::StatusCode::kOk) << status.ToString();
+
+  EXPECT_EQ(db_source.Put(key, put_value).code(), absl::StatusCode::kOk);
+
+  status = db_source.Commit();
+  EXPECT_EQ(status.code(), absl::StatusCode::kOk) << status.ToString();
+
+  // Test: Get value succeed.
+  status = db_source.Begin();
+  ASSERT_EQ(status.code(), absl::StatusCode::kOk) << status.ToString();
+
+  int64_t got_value;
+  EXPECT_EQ(db_source.Get(key, got_value).code(), absl::StatusCode::kOk);
+
+  status = db_source.Commit();
+  EXPECT_EQ(status.code(), absl::StatusCode::kOk) << status.ToString();
+  EXPECT_EQ(got_value, put_value);
+}
+
+TEST(LMDBDataSourceTest, ParallelTxnsPutAndGet) {
```

Comment:
Can you try logging each line to see where it's taking so long? It shouldn't take 5 minutes for this test to run.

Reply:
Found the root cause of timeout. It's because LMDB doesn't support multiple concurrent read-write transactions. There can be 1 read-write transaction and multiple read-only ones. Please check the updated unit test for a detailed example.